packages/hash/api/src/graphql/type-defs/knowledge/entity.typedef.ts

```diff
@@ -57,6 +58,20 @@ export const entityTypedef = gql`
       hasRightEntity: EdgeResolveDepthsInput!
     ): Subgraph!
 
+    """
+    Implementation of the Block Protocol aggregateEntities hook
+    """
+    aggregateEntities(
+      operation: AggregateOperationInput!
+      constrainsValuesOn: OutgoingEdgeResolveDepthInput!
+      constrainsPropertiesOn: OutgoingEdgeResolveDepthInput!
+      constrainsLinksOn: OutgoingEdgeResolveDepthInput!
+      constrainsLinkDestinationsOn: OutgoingEdgeResolveDepthInput!
+      isOfType: OutgoingEdgeResolveDepthInput!
+      hasLeftEntity: EdgeResolveDepthsInput!
+      hasRightEntity: EdgeResolveDepthsInput!
```

Comment:
I think it's just so we don't have to add all of the combinations. Until we support incoming depths on things like "isOfType" we won't have all of the depths on all of the queries. `getAllLatestPropertyTypes` will only take `constrainsValuesOn` and `constrainsPropertiesOn` for example

Reply:
I mean, why don't we nest all of these arguments under a `graphResolveDepths` field, as we do with `getEntity`?

It's very noisy having them at the top level, and would be easier to parse the request's arguments if you had the context of "this is all `graphResolveDepths` stuff" namespacing them

Reply:
We don't do that within `getEntity`?

```graphql
    """
    Get a subgraph rooted at an entity resolved by its id.
    """
    getEntity(
      """
      The id of the entity.
      """
      entityId: EntityId!
      """
      The version of the entity. Defaults to the latest version.
      """
      entityVersion: String
      constrainsValuesOn: OutgoingEdgeResolveDepthInput!
      constrainsPropertiesOn: OutgoingEdgeResolveDepthInput!
      constrainsLinksOn: OutgoingEdgeResolveDepthInput!
      constrainsLinkDestinationsOn: OutgoingEdgeResolveDepthInput!
      isOfType: OutgoingEdgeResolveDepthInput!
      hasLeftEntity: EdgeResolveDepthsInput!
      hasRightEntity: EdgeResolveDepthsInput!
    ): Subgraph!
```

Reply:
This would introduce _in_consistencies with the other typedefs though. Because as I mentioned, for non-entity queries the selection of depths are different.

So in GraphQL we would have to create multiple objects such as `DataTypeQueriesResolveDepths` and `PropertyTypeQueriesResolveDepths`, etc.

At the time these queries were being written, this seemed to be a lesser evil

Reply:
>  we would have to create multiple objects such as DataTypeQueriesResolveDepths and PropertyTypeQueriesResolveDepths, etc.

This seems better to me. it seems more natural / ergonomic to nest arguments if you have 7 of them related to one thing rather than spread them out. this is what we've done elsewhere with metadata and everything else.

I guess we can translate between the BP and HASH (since we have to do that for `getEntity` anyway) if it's better

Reply:
> this is what we've done elsewhere with metadata and everything else.

Yes but the metadata and such is consistent. We would be creating N permutations of the same object with just varying levels of properties here and GraphQL doesn't allow us to benefit from type composition at all.

> I guess we can translate between the BP and HASH

We are doing that already, no? I think you're just commenting on the GraphQL signature here, which is completely distinct to the signature of the hook, which _does_ take in the depths grouped in an object (`graphResolveDepths`)

Reply:
It's sunk cost if we do allow incoming edges on types for example. (Because then you could explore _from_ ontology types _to_ entities and thus you'd need all the edges and we'd have one consistent unified type.

It's also just polluting the space with a bunch of confusingly named types that all look somewhat similar to each other.

If you feel strongly that we should do this I'll put in the 15 minutes it takes to do so, just feels a bit redundant if we're hoping to get rid of most of the GraphQL API with the reintroduction of a collab service

Reply:
> If you feel strongly that we should do this I'll put in the 15 minutes it takes to do so, just feels a bit redundant if we're hoping to get rid of most of the GraphQL API with the reintroduction of a collab service

On the basis that these aren't sticking around, I agree it would be redundant. I suspect we'll have them for a good while – it's not clear to me that the only public API is going to be a collab service. Let's leave it and see if someone else suggests consolidating them at some point.

> It's also just polluting the space with a bunch of confusingly named types that all look somewhat similar to each other.

They would be the least of it, we're in the process of introducing more 😁

Reply:
> we're in the process of introducing more 😁

Let's just slap more on top then ey 😂